src/main/navigation/navigation_wind_estimator.c

```diff
@@ -0,0 +1,181 @@
+        !posControl.flags.hasValidPositionSensor ||
+        !gpsSol.flags.validVelNE ||
+        !gpsSol.flags.validVelD) {
+        return;
+    }
+
+    float groundVelocity[XYZ_AXIS_COUNT];
+    float groundVelocityDiff[XYZ_AXIS_COUNT];
+    float groundVelocitySum[XYZ_AXIS_COUNT];
+
+    float fuselageDirection[XYZ_AXIS_COUNT];
+    float fuselageDirectionDiff[XYZ_AXIS_COUNT];
+    float fuselageDirectionSum[XYZ_AXIS_COUNT];
+
+    groundVelocity[X] = gpsSol.velNED[X] / 100;
+    groundVelocity[Y] = gpsSol.velNED[Y] / 100;
+    groundVelocity[Z] = gpsSol.velNED[Z] / 100;
+
+    // TODO: Can get this from the nav system without exposing rMat?
+    fuselageDirection[X] = rMat[0][0];
+    fuselageDirection[Y] = rMat[1][0];
+    fuselageDirection[Z] = rMat[2][0];
+
+    timeDelta_t timeDelta = cmpTimeUs(currentTimeUs, lastUpdateUs);
+    // scrap our data and start over if we're taking too long to get a direction change
+    if (lastUpdateUs == 0 || timeDelta > 2 * USECS_PER_SEC) {
+
+        lastUpdateUs = currentTimeUs;
+
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+        return;
+    }
+
+    fuselageDirectionDiff[X] = fuselageDirection[X] - lastFuselageDirection[X];
+    fuselageDirectionDiff[Y] = fuselageDirection[Y] - lastFuselageDirection[Y];
+    fuselageDirectionDiff[Z] = fuselageDirection[Z] - lastFuselageDirection[Z];
+
+    float diff_length = sqrtf(sq(fuselageDirectionDiff[X]) + sq(fuselageDirectionDiff[Y]) + sq(fuselageDirectionDiff[Z]));
+    // TODO: Is 0.2f an adequate threshold?
+    if (diff_length > 0.2f) {
+        // when turning, use the attitude response to estimate wind speed
+        groundVelocityDiff[X] = groundVelocity[X] - lastGroundVelocity[X];
+        groundVelocityDiff[Y] = groundVelocity[Y] - lastGroundVelocity[Y];
+        groundVelocityDiff[Z] = groundVelocity[X] - lastGroundVelocity[Z];
+
+        // estimate airspeed it using equation 6
+        float V = (sqrtf(sq(groundVelocityDiff[0]) + sq(groundVelocityDiff[1]) + sq(groundVelocityDiff[2]))) / diff_length;
+
+        fuselageDirectionSum[X] = fuselageDirection[X] + lastFuselageDirection[X];
+        fuselageDirectionSum[Y] = fuselageDirection[Y] + lastFuselageDirection[Y];
+        fuselageDirectionSum[Z] = fuselageDirection[Z] + lastFuselageDirection[Z];
+
+        groundVelocitySum[X] = groundVelocity[X] + lastGroundVelocity[X];
+        groundVelocitySum[Y] = groundVelocity[Y] + lastGroundVelocity[Y];
+        groundVelocitySum[Z] = groundVelocity[Z] + lastGroundVelocity[Z];
+
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+
+        float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(groundVelocityDiff[1], groundVelocityDiff[0]);// equation 9
```

Comment:
We need to use fastmath functions here: `atan2_approx`, `sin_approx` and `cos_approx`.